File: baekjoon/04-DP/python/11053.py
# https://www.acmicpc.net/problem/11053 가장 긴 증가하는 부분 수열 (ms)
import sys
sys.stdin = open('input.txt', 'r')
input = sys.stdin.readline

n = int(input())
a = list(map(int, input().split()))
dp = [1] * n
# n의 크기가 1e3 이기 때문에 dp로 풀어도 됨
for i in range(n):
    for j in range(i):
        if a[i] > a[j]: 
            dp[i] = max(dp[i], dp[j]+1)

print(max(dp))

File: baekjoon/04-DP/js/11053.js
// 백준 제출 시, readFileSync('/dev/stdin') 으로 변경할 것
const fs = require("fs");
const input = fs.readFileSync("예제.txt").toString().split("\n");

const n = Number(input[0]);
const arr = input[1].split(" ").map(Number);
let dp = new Array(n).fill(1);

for (let i = 0; i < n; i++) {
    for (let j = 0; j < i; j++) {
        if (arr[i] > arr[j]) {
            dp[i] = Math.max(dp[i], dp[j] + 1);
        }
    }
}

console.log(Math.max(...dp));
